File: quant-backend/line_notifier.py
import os
import json
import urllib.request
import ssl
import logging

# LINEサーバー側の.envパスを特定
LINE_ENV_PATH = "/Users/ooshirokazuki2/.gemini/antigravity/scratch/line-agent-controller/.env"

logger = logging.getLogger(__name__)

def send_line_push(text):
    """理系専務から社長へのLINE通知実行"""
    try:
        # .envの読み込み
        env = {}
        if not os.path.exists(LINE_ENV_PATH):
            logger.error(f"LINE Config Not Found at {LINE_ENV_PATH}")
            return False

        with open(LINE_ENV_PATH) as f:
            for line in f:
                if line.strip() and not line.startswith('#'):
                    k, v = line.strip().split('=', 1)
                    env[k] = v.strip('"\'')

        access_token = env.get('LINE_CHANNEL_ACCESS_TOKEN', '')
        user_id = env.get('LINE_ADMIN_USER_ID', '')

        if not access_token or not user_id:
            logger.error("LINE Credentials Missing in .env")
            return False

        url = 'https://api.line.me/v2/bot/message/push'
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f"Bearer {access_token}"
        }
        data = {
            'to': user_id,
            'messages': [{'type': 'text', 'text': text}]
        }

        # SSL / Mac対応
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE

        req = urllib.request.Request(url, data=json.dumps(data).encode('utf-8'), headers=headers)
        urllib.request.urlopen(req, context=ctx)
        logger.info("✅ LINE通知を送信しました")
        return True

    except Exception as e:
        logger.error(f"LINE通知エラー: {e}")
        return False

if __name__ == "__main__":
    # テスト用
    send_line_push("【理系専務 起動テスト】\n社長、バックエンド側の通知システムが正常にリンクされました。これより市場の自己監査結果を随時報告いたします。")

File: line-agent-controller/lib/line_notifier.ts
// LINE Developer Consoleから取得した情報を環境変数に設定します
const LINE_CHANNEL_ACCESS_TOKEN = process.env.LINE_CHANNEL_ACCESS_TOKEN || '';
// セキュリティのため、あなたのユーザーIDにのみ送信を許可する設定
const ADMIN_USER_ID = process.env.LINE_ADMIN_USER_ID || '';

/**
 * 緊急アラートや承認依頼をあなた（社長）にプッシュ送信する関数
 * ※これは月200通の無料枠を消費する通信です。
 */
export async function sendAdminAlert(message: string, requireApproval: boolean = false) {
  if (!ADMIN_USER_ID) {
    console.error('管理者のLINE IDが設定されていません。通知スキップ。');
    return;
  }

  const messagesPayload: any[] = [];

  if (requireApproval) {
    // 承認が必要な場合は、「Quick Reply」を使ってボタンを表示させる
    messagesPayload.push({
      type: 'text',
      text: `⚠️ 【緊急確認】\n${message}`,
      quickReply: {
        items: [
          {
            type: 'action',
            action: {
              type: 'message',
              label: '👍 承認して続行',
              text: '承認'
            }
          },
          {
            type: 'action',
            action: {
              type: 'message',
              label: '✖️ 却下して中止',
              text: '却下'
            }
          }
        ]
      }
    });
  } else {
    // 単なるテキスト通知の場合
    messagesPayload.push({
      type: 'text',
      text: `ℹ️ 【システム報告】\n${message}`
    });
  }

  try {
    const res = await fetch('https://api.line.me/v2/bot/message/push', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${LINE_CHANNEL_ACCESS_TOKEN}`
      },
      body: JSON.stringify({
        to: ADMIN_USER_ID, // 宛先を「あなた」に強制・固定
        messages: messagesPayload
      })
    });

    if (!res.ok) {
      console.error('プッシュ通知エラー:', await res.text());
    } else {
      console.log('✅ LINEへプッシュ通知を送信しました。');
    }
  } catch (error) {
    console.error('送信用APIエラー:', error);
  }
}
